### ISSUE TITLE
SCIM Outbound Provisioning always Connection Refused (?)

### ISSUE THREAD

**Issue by kiwinesian** [wso2/product-is#12213]
**Describe the issue:**
Using the SCIM Outbound provisioning always led to `Connection refused` to the endpoint.
Endpoint can be `GET` and `POST` via curl and postman. with the exact same configuration

```
ERROR {org.wso2.carbon.identity.provisioning.ProvisioningThread} -  Provisioning for Entity testuser For operation = POST org.wso2.carbon.identity.provisioning.IdentityProvisioningException: Error while creating the user
        at org.wso2.carbon.identity.provisioning.connector.scim.SCIMProvisioningConnector.createUser(SCIMProvisioningConnector.java:281)
        at org.wso2.carbon.identity.provisioning.connector.scim.SCIMProvisioningConnector.provision(SCIMProvisioningConnector.java:109)
        at org.wso2.carbon.identity.provisioning.ProvisioningThread.call(ProvisioningThread.java:68)
        at org.wso2.carbon.identity.provisioning.ProvisioningThread.call(ProvisioningThread.java:32)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.wso2.carbon.identity.scim.common.utils.IdentitySCIMException: Error in invoking provisioning operation for the user with id: gwadmin
        at org.wso2.carbon.identity.scim.common.impl.ProvisioningClient.provisionCreateUser(ProvisioningClient.java:172)
        at org.wso2.carbon.identity.provisioning.connector.scim.SCIMProvisioningConnector.createUser(SCIMProvisioningConnector.java:257)
        ... 7 more
Caused by: java.net.ConnectException: Connection refused (Connection refused)
        at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
        at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
        at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
        at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
        at java.base/java.net.Socket.connect(Socket.java:609)
        at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:289)
        at java.base/sun.security.ssl.SSLSocketImpl.<init>(SSLSocketImpl.java:194)
        at java.base/sun.security.ssl.SSLSocketFactoryImpl.createSocket(SSLSocketFactoryImpl.java:153)
        at org.apache.commons.httpclient.protocol.SSLProtocolSocketFactory.createSocket(SSLProtocolSocketFactory.java:172)
        at org.apache.commons.httpclient.HttpConnection.open(HttpConnection.java:707)
        at org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.java:387)
        at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:171)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:397)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:323)
        at org.wso2.carbon.identity.scim.common.impl.ProvisioningClient.provisionCreateUser(ProvisioningClient.java:146)
```

**How to reproduce:**
1. Create a new IdP
2. Set the Outbound Provisioning to the endpoint
3. Enable the IdP Outbound Provisioning at the `Resident Service Provider`
4. Add a user
5. Check the log file

**Expected behavior:**
User should be provisioned to the targeted endpoint

**Environment information** (_Please complete the following information; remove  any unnecessary fields_) **:**
 - Product Version: 5.11
 - OS: Docker
 - Database: H2
 - Userstore: 

---

### Optional Fields

**Configuration screenshots**

#### WSO2 IS 5.11 Configuration
![image](https://user-images.githubusercontent.com/12476195/127750394-6d8ed932-15f8-4eb1-8e4f-da6c9dc97c61.png)

#### Query via Postman
![image (1)](https://user-images.githubusercontent.com/12476195/127750402-ecf079f3-22de-4fb8-9906-2a6250443067.png)

#### Query via `curl`
![image (2)](https://user-images.githubusercontent.com/12476195/127750409-b86cfaea-3ca0-4c65-95ab-055501105215.png)

**Comment by kiwinesian** [comment]
hmmm I just try between Identity Server to Identity server and still the same issue? 
One is listening at port 9443 and the other is listening at port 9543

#### Docker images
![image](https://user-images.githubusercontent.com/12476195/127755627-a448dd4d-3464-405c-a84f-2da50f9e3e54.png)

#### Log file
![image](https://user-images.githubusercontent.com/12476195/127755645-52885d7e-2554-452a-8bf4-eb85616202e2.png)

**Comment by kiwinesian** [comment]
going the IS2 to IS1 gives an error indicating duplicate even though there isn't any duplicate.
What is going on here 🤔 

![image](https://user-images.githubusercontent.com/12476195/127755737-a628c83c-cc0f-4596-9c0a-a9d3abd6240e.png)

**Comment by Amlesh-NEC** [comment]
Hi @kiwinesian,

As I analysis this issue, I believe multiple possible ways to throw this exception:

1. Client and Server, either or both of them are not in the network: LAN or internet or any other network, in that case, Java will throw "java.net.ConnectException: Connection refused" exception on client side.
2. server is down and not running. In that case, also you will get java.net.ConnectException: Connection refused error.
3. server is running but listening on a different port: cause of "java.net.ConnectException: Connection refused".

**Comment by kiwinesian** [comment]
hiiii thanks for looking into this!

They are both in the same localhost, but is running on docker.
each docker listening on different port, 9443:9443 and 9543:9443

I also tried to change the application port to match as well, 9443:9443 and
9543:9543, but still getting connection refused.

i think the `localhost` domain is not whitelisted properly?

I finally managed to get it going by deploying it to our VM and front it
with NGINX TLS certs from Lets Encrypt.
